_posts/2022-05-12-faster-rand.md

```diff
@@ -0,0 +1,1096 @@
+L2272:
+    mov qword ptr [rbx+8], rax
+# line_I
+```
+Here is `band` with similar test and fallback code:
+```nasm
+# i_band_ssjd
+    mov rsi, qword ptr [rbx]
+    mov rax, 68719476735
+# is the operand small?
+    mov edi, esi
+    and edi, 15
+    cmp edi, 15
+    short jne L2273
+    and rax, rsi
+    short jmp L2274
+L2273:
+    call 140439031216768
+L2274:
+    mov qword ptr [rbx], rax
+```
+Below comes `*` with test, fallback code,
+and overflow check:
+```nasm
+# line_I
+# i_times_jssd
+    mov rsi, qword ptr [rbx]
+    mov edx, 2141605935
+# is the operand small?
+    mov edi, esi
+    and edi, 15
+    cmp edi, 15
+    short jne L2276
+# mul with overflow check, imm RHS
+    mov rax, rsi
+    mov rcx, 133850370
+    and rax, -16
+    imul rax, rcx
+    short jo L2276
+    or rax, 15
+    short jmp L2275
+L2276:
+    call 140439031220000
+L2275:
+    mov qword ptr [rbx], rax
+```
+The following is `+` with tests, fallback code, and overflow check:
+```nasm
+# i_plus_ssjd
+    mov rsi, qword ptr [rbx]
+    mov rdx, qword ptr [rbx+8]
+# are both operands small?
+    mov eax, esi
+    and eax, edx
+    and al, 15
+    cmp al, 15
+    short jne L2278
+# add with overflow check
+    mov rax, rsi
+    mov rcx, rdx
+    and rcx, -16
+    add rax, rcx
+    short jno L2277
+L2278:
+    call 140439031219296
+L2277:
+    mov qword ptr [rbx], rax
+```
+
+### With type-based optimization
+
+When the compiler can figure out type information about the arguments
+it can emit more efficient code.  One would like to add a guard
+that restricts the argument to a 59 bit integer, but unfortunately
+the compiler cannot yet make use of such a guard test.
+
+But adding a redundant input bit mask to the Erlang code puts the compiler
+on the right track.  This is a kludge, and will only be used
+until the compiler has been improved to deduce the same information
+from a guard instead.
+
+The Erlang code now has a first redundant mask to 59 bits:
+``` erlang
+mwc59(CX0) ->
+    CX = CX0 band ((1 bsl 59)-1),
+    C = CX band ((1 bsl 32)-1),
+    X = CX bsr 32,
+    16#7fa6502 * X + C.
+```
+The BEAM assembler then becomes, with the default type-based optimizations
+in the compiler the OTP-25.0 release:
+``` erlang
+    {gc_bif,'band',{f,0},1,[{x,0},{integer,576460752303423487}],{x,0}}.
+    {gc_bif,'bsr',
+            {f,0},
+            1,
+            [{tr,{x,0},{t_integer,{0,576460752303423487}}},{integer,32}],
+            {x,1}}.
+    {gc_bif,'band',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,{0,576460752303423487}}},
+             {integer,4294967295}],
+            {x,0}}.
+    {gc_bif,'*',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,{0,4294967295}}},{integer,133850370}],
+            {x,0}}.
+    {gc_bif,'+',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,{0,572367635452168875}}},
+             {tr,{x,1},{t_integer,{0,134217727}}}],
+            {x,0}}.
+```
```

Comment:
```suggestion
    {gc_bif,'band',{f,0},1,[{x,0},{integer,576460752303423487}],{x,0}}.
    {gc_bif,'bsr',{f,0},1,[{tr,{x,0},{t_integer,{0,576460752303423487}}},
             {integer,32}],{x,1}}.
    {gc_bif,'band',{f,0},2,[{tr,{x,0},{t_integer,{0,576460752303423487}}},
             {integer,4294967295}],{x,0}}.
    {gc_bif,'*',{f,0},2,[{tr,{x,0},{t_integer,{0,4294967295}}},
             {integer,133850370}],{x,0}}.
    {gc_bif,'+',{f,0},2,[{tr,{x,0},{t_integer,{0,572367635452168875}}},
             {tr,{x,1},{t_integer,{0,134217727}}}],{x,0}}.
```
```

Maybe collapsing this will make it more readable? or maybe not...

Reply:
Cut and paste from the shell.  I agree reformatting makes it more readable.